Add Search filters to Notes Grid

andres-holguin opened the pull request:
# Search Results Filter
This PR should close #20
The ability to filter search results has been added.
![Starfish-Filter-List-V1](https://user-images.githubusercontent.com/44207690/88007896-f6c5de80-cadc-11ea-9965-943ba5869ce1.gif)

## Main Features
- A new FilterList component that is inside the NoteGrid component.
  - Has checkboxes with predetermined set of Common MiscLabels to filter by.
  - Updates a combobox where user can enter their own filters.
- A helper component: MockSearchButton.vue
  - Acts like the SearchBar--will emit an array of sample NoteObjects which can be used for testing the UI without a backend.

## Todo:
- Enhance Styling
- Modify FilterList to fetch suggestedLabels when School or Course changes.
- Create backend servlet that serves common MiscLabels given School/Course
- Modify SearchBar to emit current query's School and Course

andres-holguin committed:
Create MockSearchButton.vue

Create a component that acts like the searchbar, emitting mock note data for UI testing.

andres-holguin committed:
Update NoteGrid.vue

Remove debug message

andres-holguin committed:
Update NoteGrid.vue

Add filter data and include as slot prop

andres-holguin committed:
Update NoteGridCollection.vue

Include functionality for filtering a NoteGridCollection's notes.

andres-holguin committed:
Update MockSearchButton and Test.html

Update for testing filter capability.

andres-holguin committed:
Update test.html

Use slot props for note-grid filter.

andres-holguin committed:
Update NoteGridCollection.vue

- Abstract the filter function to noteFilter
- Expand filter to incldude note title, school, and course

andres-holguin committed:
Update MockSearchButton.vue

Further specify mock data

andres-holguin committed:
Create FilterList.vue

Create component with filter selection for search results.

andres-holguin committed:
Update NoteGrid with FilterList

Add the FilterList component to the NoteGrid

andres-holguin committed:
Update FilterList.vue to fetch labels

The FilterList will fetch the suggested labels based on the School and Course used in the search query.

keelywan commented on src/main/webapp/components/FilterList.vue:
```diff
@@ -0,0 +1,66 @@
+  },
+  methods: {
+    fetchSuggestedLabels: function() {
+      let url = new URL("/common-labels", window.location.href);
```
Is this the servlet for retrieving all the school and course labels? Or is this something else?

Aradhya-B replied:
If it is a call to the servlet for retrieving all the school and course labels, I named the path `/get-school-and-course-labels` but we could also rename it on #57 if needed.

andres-holguin replied:
Oops, forgot a TODO. This is a servlet that does not exist yet haha. It should serve some of the most common labels for each school/course.
```suggestion
      let url = new URL("/common-labels", window.location.href); // TODO: Create the servlet and rename "/common-labels"
```

Aradhya-B replied:
ohhhh gotcha, then ignore my comment on aligning with the servlet in #57!

Aradhya-B replied:
I can go ahead and create the `/common-labels` servlet today, I'll add it to the project board.

Aradhya-B replied:
Now that I think about it though, I'm not sure we'd be able to create a `/common-labels` servlet without making a modification to the schema to track the number of times any given label has been searched for. Like each label is only 1 row in the Labels table at this point so I think to get the 10 most common for example, we'd need to add a column on the Labels table `numTimesSearched` or something that is basically just a counter that goes up by 1 everytime that label is searched for. Then we could simply query for labels ordered descending by `numTimesSearched`.
I could definitely work on this but I'm not sure if this is something that we want to include in our scope for this week or whether this should be considered a moonshot. If it isn't something we want to get done immediately, then it'd be good to make changes here to reflect that and assume we don't have a `/common-labels` path.

andres-holguin replied:
I was thinking the most common labels assigned to the notes on the DB, not searched for. So something like
```
SELECT label FROM
  miscLabels INNER JOIN (
    SELECT noteID FROM notes
    WHERE course="course" AND school="school"
  ) 
GROUP BY label
SORT BY count(label) DESC
LIMIT 10
 ```
I'm sure this will require a lot of tweaking, but no need to change the schema, I think.

keelywan reviewed (approved):
This looks really good!

andres-holguin committed:
Add TODO to FilterList.vue

Mention that servlet does not exist yet.

Aradhya-B commented on src/main/webapp/components/NoteGridCollection.vue:
```diff
@@ -18,11 +18,31 @@ module.exports = {
+  computed: {
+    activeNotes: function() {
+      // Filter the notes that get shown based on filter provided by parent.
+      if (this.filters && this.filters.length) {
+        return this.notes.filter(this.noteFilter);
+      } else {
+        return this.notes;
+      }
```
nit: Don't need `else` if you're returning from inside the `if`.
```suggestion
      if (this.filters && this.filters.length) {
        return this.notes.filter(this.noteFilter);
      }
      return this.notes;
```

andres-holguin replied:
Thanks for the suggestion :)

Aradhya-B commented on src/main/webapp/test.html:
```diff
@@ -1,30 +1,62 @@
+    <link rel="stylesheet" href="edit-profile.css">
+    <script src="edit-profile.js"></script>
+    <link rel="stylesheet" href="styles/landingpage.css">
+    <!-- <script src="script.js"></script> -->
```
Do we still need this commented out code?

andres-holguin replied:
No, that's my bad, this file shouldn't have even been committed. It's basically just a sandbox for debugging, shouldn't be on master.

Aradhya-B commented on src/main/webapp/components/FilterList.vue:
```diff
@@ -0,0 +1,66 @@
+      url.searchParams.set('school', this.school);
+      url.searchParams.set('course', this.course);
+
+      fetch(url)
+        .then(response => response.json())
+        .then(result => {
+          this.suggestedFilters = result;
+        })
```
If this is a call to the backend to get all the school and course labels, the way I ended up designing the API was that you'd get back an object where the two keys are `"schools"` and `"courses"` and their values are String arrays of labels of those type. So you could either just combine those two arrays and make that the `suggestedFilters` or you could keep two variables `suggestedSchoolFilters` and `suggestedCourseFilters` for the two arrays; I'm not sure which of these options is more useful to you on the frontend. Feel free to add a review comment on #57 if you think that API should respond differently than it does.

Aradhya-B reviewed (changes requested):
Looks really good, great work!!

andres-holguin commented:
> We found a Contributor License Agreement for you (the sender of this pull request), but were unable to find agreements for all the commit author(s) or Co-authors. If you authored these, maybe you used a different email address in the git commits than was used to sign the CLA ([login here](https://cla.developers.google.com/) to double check)? If these were authored by someone else, then they will need to sign a CLA as well, and confirm that they're okay with these being contributed to Google.
> In order to pass this check, please resolve this problem and then comment `@googlebot I fixed it.`. If the bot doesn't comment, it means it doesn't think anything has changed.
> 
> ℹ️ **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2Fgoogleinterns%2Fstep102-2020%2Fpull%2F58) for more info**.

@Aradhya-B would you mind signing the CLA for your aradhyabansal2000@gmail.com email? If not I can probably just force push to remove that last commit.

andres-holguin committed:
Tweak NoteGridCollection.vue

Remove unecessary else statement.

andres-holguin committed:
Add sortFunc prop to NoteGridCollection

Add a prop to NoteGridCollection which is used to compare notes for sorting the collection.

andres-holguin committed:
Rename prop in NoteGridCollection

Rename sortFunc to compareFunc, and propogate that change.

andres-holguin committed:
Update NotePreview.vue to remove computed src

As per a PR discussion, the computed src is removed since the back end will provide a url as the pdfSrc

andres-holguin committed:
Update MockSearchButton with unique Dates

Useful for testing the sort by time feature.

andres-holguin committed:
Update MockSearchButton with new dates

Use dates relative to current time rather than Date(0)
